File: benchmarks/tsb/bench_read_sas.ts
/**
 * Benchmark: readSas — parse a 1,000-row SAS XPORT (XPT) file.
 * Outputs JSON: {"function": "read_sas", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { readSas } from "../../src/index.ts";

// ─── IBM 370 floating-point encoder ──────────────────────────────────────────

function ibmEncode(val: number): Uint8Array {
  const out = new Uint8Array(8);
  if (val === 0) return out;
  if (!Number.isFinite(val)) { out[0] = 0x2e; return out; }
  const sign = val < 0 ? 1 : 0;
  const abs = Math.abs(val);
  let exp = 0;
  let mant = abs;
  while (mant >= 1) { mant /= 16; exp++; }
  while (mant < 1 / 16 && mant > 0) { mant *= 16; exp--; }
  const mantInt = BigInt(Math.round(mant * 2 ** 56));
  out[0] = (sign << 7) | ((exp + 64) & 0x7f);
  for (let i = 1; i <= 7; i++) {
    out[i] = Number((mantInt >> BigInt((7 - i) * 8)) & 0xffn);
  }
  return out;
}

// ─── Minimal XPORT v5 builder ────────────────────────────────────────────────

function buildXpt(
  numVars: readonly string[],
  charVars: readonly { name: string; len: number }[],
  rows: readonly Readonly<Record<string, number | string>>[],
): Uint8Array {
  const RECORD = 80;

  function encodeAscii(s: string, maxLen: number): Uint8Array {
    const buf = new Uint8Array(maxLen);
    for (let i = 0; i < Math.min(s.length, maxLen); i++) buf[i] = s.charCodeAt(i) & 0x7f;
    return buf;
  }
  function padTo80(s: string): Uint8Array { return encodeAscii(s.padEnd(RECORD, " "), RECORD); }
  function writeU16(b: Uint8Array, o: number, v: number) { b[o] = (v >> 8) & 0xff; b[o + 1] = v & 0xff; }
  function writeU32(b: Uint8Array, o: number, v: number) {
    b[o] = (v >> 24) & 0xff; b[o + 1] = (v >> 16) & 0xff; b[o + 2] = (v >> 8) & 0xff; b[o + 3] = v & 0xff;
  }

  type Meta = { type: 1 | 2; name: string; len: number; pos: number };
  const metas: Meta[] = [];
  let pos = 0;
  for (const name of numVars) { metas.push({ type: 1, name, len: 8, pos }); pos += 8; }
  for (const { name, len } of charVars) { metas.push({ type: 2, name, len, pos }); pos += len; }
  const rowLen = pos;

  const chunks: Uint8Array[] = [];

  // Library header (5 × 80 bytes)
  chunks.push(padTo80("HEADER RECORD*******LIBRARY HEADER RECORD!!!!!!!000000000000000000000000000000  "));
  chunks.push(padTo80("SAS     SAS     SASLIB  6.06    ASCII"));
  chunks.push(padTo80("20240101"));
  chunks.push(padTo80(""));
  chunks.push(padTo80(""));

  // Member header (3 × 80 bytes)
  chunks.push(padTo80("HEADER RECORD*******MEMBER  HEADER RECORD!!!!!!!000000000000000000000000000001600000000140  "));
  chunks.push(padTo80("SAS     BENCH   SASDATA 6.06    ASCII"));
  chunks.push(padTo80(""));

  // Namestr header
  const nvar = metas.length;
  chunks.push(padTo80(`HEADER RECORD*******NAMESTR HEADER RECORD!!!!!!!${String(nvar).padStart(6, "0")}00000000000000000000  `));

  // Namestr records (140 bytes each)
  const nsBuf = new Uint8Array(nvar * 140);
  for (let i = 0; i < metas.length; i++) {
    const m = metas[i]!;
    const off = i * 140;
    writeU16(nsBuf, off, m.type);
    writeU16(nsBuf, off + 2, 140);
    nsBuf.set(encodeAscii(m.name, 8), off + 4);
    writeU16(nsBuf, off + 52, m.len);
    writeU32(nsBuf, off + 84, m.pos);
  }
  const nsPadded = Math.ceil(nsBuf.length / RECORD) * RECORD;
  const nsPaddedBuf = new Uint8Array(nsPadded);
  nsPaddedBuf.set(nsBuf);
  chunks.push(nsPaddedBuf);

  // Obs header
  chunks.push(padTo80("HEADER RECORD*******OBS     HEADER RECORD!!!!!!!000000000000000000000000000000  "));

  // Observations
  const paddedRowLen = Math.ceil(rowLen / RECORD) * RECORD;
  const obsBuf = new Uint8Array(rows.length * paddedRowLen);
  for (let r = 0; r < rows.length; r++) {
    const base = r * paddedRowLen;
    const row = rows[r]!;
    for (const m of metas) {
      const val = row[m.name];
      if (m.type === 1) {
        const encoded = ibmEncode(typeof val === "number" ? val : 0);
        obsBuf.set(encoded, base + m.pos);
      } else {
        const s = typeof val === "string" ? val : "";
        obsBuf.set(encodeAscii(s, m.len), base + m.pos);
      }
    }
  }
  chunks.push(obsBuf);

  let total = 0;
  for (const c of chunks) total += c.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) { out.set(c, off); off += c.length; }
  return out;
}

// ─── Build dataset ────────────────────────────────────────────────────────────

const ROWS = 1_000;
const WARMUP = 3;
const ITERATIONS = 20;

const rows: Readonly<Record<string, number | string>>[] = Array.from({ length: ROWS }, (_, i) => ({
  id: i,
  value: i * 1.5,
  score: Math.sin(i * 0.01),
  label: `item_${i % 100}`,
}));

const xpt = buildXpt(["id", "value", "score"], [{ name: "label", len: 12 }], rows);

// ─── Benchmark ────────────────────────────────────────────────────────────────

for (let i = 0; i < WARMUP; i++) readSas(xpt);

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) readSas(xpt);
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "read_sas",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_read_sas.py
"""
Benchmark: pd.read_sas — parse a 1,000-row SAS XPORT (XPT) file.
Outputs JSON: {"function": "read_sas", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import io
import struct
import math
import numpy as np
import pandas as pd

ROWS = 1_000
WARMUP = 3
ITERATIONS = 20

# ── IBM 370 double encoder ────────────────────────────────────────────────────

def ibm_encode(val: float) -> bytes:
    if val == 0.0:
        return b"\x00" * 8
    if not math.isfinite(val):
        return b"\x2e" + b"\x00" * 7
    sign = 1 if val < 0 else 0
    abs_val = abs(val)
    exp = 0
    mant = abs_val
    while mant >= 1.0:
        mant /= 16
        exp += 1
    while mant < 1.0 / 16 and mant > 0:
        mant *= 16
        exp -= 1
    mant_int = round(mant * 2**56)
    out = bytearray(8)
    out[0] = (sign << 7) | ((exp + 64) & 0x7f)
    for i in range(1, 8):
        out[i] = (mant_int >> ((7 - i) * 8)) & 0xff
    return bytes(out)

# ── Minimal XPORT v5 builder ─────────────────────────────────────────────────

def build_xpt(num_vars, char_vars, rows_data):
    RECORD = 80

    def pad80(s):
        return s.ljust(RECORD).encode("ascii")[:RECORD]

    def write_u16(val):
        return struct.pack(">H", val)

    def write_u32(val):
        return struct.pack(">I", val)

    # Compute variable metadata
    metas = []
    pos = 0
    for name in num_vars:
        metas.append({"type": 1, "name": name, "len": 8, "pos": pos})
        pos += 8
    for name, length in char_vars:
        metas.append({"type": 2, "name": name, "len": length, "pos": pos})
        pos += length
    row_len = pos

    chunks = bytearray()

    # Library header (5 × 80 bytes)
    chunks += pad80("HEADER RECORD*******LIBRARY HEADER RECORD!!!!!!!000000000000000000000000000000  ")
    chunks += pad80("SAS     SAS     SASLIB  6.06    ASCII")
    chunks += pad80("20240101")
    chunks += pad80("")
    chunks += pad80("")

    # Member header (3 × 80 bytes)
    chunks += pad80("HEADER RECORD*******MEMBER  HEADER RECORD!!!!!!!000000000000000000000000000001600000000140  ")
    chunks += pad80("SAS     BENCH   SASDATA 6.06    ASCII")
    chunks += pad80("")

    # Namestr header
    nvar = len(metas)
    chunks += pad80(f"HEADER RECORD*******NAMESTR HEADER RECORD!!!!!!!{nvar:06d}00000000000000000000  ")

    # Namestr records (140 bytes each)
    ns_buf = bytearray(nvar * 140)
    for i, m in enumerate(metas):
        off = i * 140
        ns_buf[off:off+2] = write_u16(m["type"])
        ns_buf[off+2:off+4] = write_u16(140)
        name_bytes = m["name"].encode("ascii").ljust(8)[:8]
        ns_buf[off+4:off+12] = name_bytes
        ns_buf[off+52:off+54] = write_u16(m["len"])
        ns_buf[off+84:off+88] = write_u32(m["pos"])
    padded_ns = math.ceil(len(ns_buf) / RECORD) * RECORD
    ns_buf_padded = ns_buf.ljust(padded_ns, b"\x00")
    chunks += ns_buf_padded

    # Obs header
    chunks += pad80("HEADER RECORD*******OBS     HEADER RECORD!!!!!!!000000000000000000000000000000  ")

    # Observations
    padded_row_len = math.ceil(row_len / RECORD) * RECORD
    obs_buf = bytearray(len(rows_data) * padded_row_len)
    for r, row in enumerate(rows_data):
        base = r * padded_row_len
        for m in metas:
            val = row.get(m["name"])
            if m["type"] == 1:
                encoded = ibm_encode(float(val) if val is not None else 0.0)
                obs_buf[base + m["pos"]:base + m["pos"] + 8] = encoded
            else:
                s = str(val) if val is not None else ""
                b = s.encode("ascii")[:m["len"]].ljust(m["len"], b" ")
                obs_buf[base + m["pos"]:base + m["pos"] + m["len"]] = b

    chunks += obs_buf
    return bytes(chunks)


# ── Build dataset ─────────────────────────────────────────────────────────────

rows_data = [
    {"id": float(i), "value": i * 1.5, "score": math.sin(i * 0.01), "label": f"item_{i % 100}"}
    for i in range(ROWS)
]

xpt_bytes = build_xpt(
    ["id", "value", "score"],
    [("label", 12)],
    rows_data,
)

# ── Benchmark ─────────────────────────────────────────────────────────────────

for _ in range(WARMUP):
    pd.read_sas(io.BytesIO(xpt_bytes), format="xport")

start = time.perf_counter()
for _ in range(ITERATIONS):
    pd.read_sas(io.BytesIO(xpt_bytes), format="xport")
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "read_sas",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
